### PR TITLE
Criteria Query task

### PR DESCRIPTION
**Pull request by Stelmaschyk**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder)
+                -> root.get("title").in(Arrays.stream(params).toArray());
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/book/spec/TitleSpecificationProvider.java:18]: extract repeated literal to a constant

File: src/main/java/mate/academy/bookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder)
+                -> root.get("author").in(Arrays.stream(params).toArray());
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/book/spec/AuthorSpecificationProvider.java:18]: same

File: src/main/java/mate/academy/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+
+    private final List<SpecificationProvider<Book>> specificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return specificationProviders.stream()
+            .filter(p -> p.getKey().equals(key)).findFirst()
+            .orElseThrow(()
+                    -> new RuntimeException("No specification provider found for key: " + key));
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/book/BookSpecificationProviderManager.java]:
```suggestion
            .filter(provider -> provider.getKey().equals(key)).findFirst()
            .orElseThrow(()
                    -> new NoSuchElementException("No specification provider found for key: " + key));
```

File: src/main/java/mate/academy/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.RequiredArgsConstructor;
+import mate.academy.bookstore.model.Book;
+import mate.academy.bookstore.repository.SpecificationProvider;
+import mate.academy.bookstore.repository.SpecificationProviderManager;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/book/BookSpecificationProviderManager.java]:
```suggestion
```
don't leave an empty line after method/class definition

File: src/main/java/mate/academy/bookstore/repository/book/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import mate.academy.bookstore.model.Book;
+import mate.academy.bookstore.repository.SpecificationBuilder;
+import mate.academy.bookstore.repository.SpecificationProviderManager;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/book/BookSpecificationBuilder.java]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**